dot/core/digest.go

```diff
@@ -0,0 +1,118 @@
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
+package core
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ChainSafe/gossamer/dot/core/types"
+	"github.com/ChainSafe/gossamer/lib/babe"
+	log "github.com/ChainSafe/log15"
+)
+
+// handleBlockDigest checks if the provided header is the block header for
+// the first block of the current epoch, finds and decodes the consensus digest
+// item from the block digest, and then sets the epoch data for the next epoch
+func (s *Service) handleBlockDigest(header *types.Header) (err error) {
+
+	// check if block header digest items exist
+	if header.Digest == nil || len(header.Digest) == 0 {
+		return fmt.Errorf("header digest is not set")
+	}
+
+	// declare digest item
+	var item types.DigestItem
+
+	// decode each digest item and check its type
+	for _, digest := range header.Digest {
+		item, err = types.DecodeDigestItem(digest)
+		if err != nil {
+			return err
+		}
+
+		// check if digest item is consensus digest type
+		if item.Type() == types.ConsensusDigestType {
+
+			// cast decoded consensus digest item to conensus digest type
+			consensusDigest, ok := item.(*types.ConsensusDigest)
+			if !ok {
+				return errors.New("failed to cast DigestItem to ConsensusDigest")
+			}
+
+			err = s.handleConsensusDigest(header, consensusDigest)
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// handleConsensusDigest checks if the first block has been set for the current
+// epoch, if first block has not been set or the block header has a lower block
+// number, sets epoch data for the next epoch using data in the ConsensusDigest
+func (s *Service) handleConsensusDigest(header *types.Header, digest *types.ConsensusDigest) error {
+
+	// check if block header is from current epoch
+	currentEpoch, err := s.blockFromCurrentEpoch(header.Hash())
+	if err != nil {
+		return fmt.Errorf("failed to check if block header is from current epoch: %s", err)
+	} else if !currentEpoch {
+		return fmt.Errorf("block header is not from current epoch: %s", err)
+	}
+
+	// check if first block has been set for current epoch
+	if s.firstBlock != nil {
+
+		// check if first block has lower block number than block header
+		if s.firstBlock.Cmp(header.Number) != -1 {
+			return fmt.Errorf("first block already set for current epoch")
+		}
+
+		// log error if BABE produced two first blocks or received false first block from connected peer
+		log.Error("[core] received first block header with lower block number than current first block")
```

Comment:
if header has lower block number than first block, log error and replace current first block with header

Reply:
maybe change this to just `Warn` or `Debug`?

Reply:
done :+1: